Fixing in bug in suggestions list visibility

anthonygore opened the pull request:
Hi Toran,

After looking at how `onInput` functions, I realised that the `suggestions` array is populated by the user's implementation of `determineSuggestions` after user input is received. I said before that you don't need `determineSuggestions` and that you should just implement `suggestions`, but I was wrong about that. So I've added it back into the documentation.

However, `onInput` needed updating. Hopefully it's self explanatory, but the user should implement `determineSuggestions` as an Ember computed property, not as a function.

I also had a use case that demonstrated a bug: if you have a really long list of suggestions, you may not want to return any suggestions until the user has typed at least 3 characters (for example), otherwise it works really slowly. Here's an example implementation of `determineSuggestions` that does that:

```
determineSuggestions: function() {
    var inputVal = this.get("inputVal") || "";
    if (inputVal.length >= 3) {
        return this.get("options").filter(function(item) {
            return item.get("label").toLowerCase().indexOf(inputVal.toLowerCase()) > -1;
        });
    }
    return [];
}.property("inputVal", "options.@each")
```

The bug is that the suggestions list stays hidden until the user has typed 4 characters (in this example...the general case is that it stays hidden until the user type x + 1 chars where x is the length of `inputVal`). Triggering `onInput` by a change in `inputVal` and triggering `setVisible` after that change fixes the problem. This means `suggestions` should be initialised as an empty array to avoid errors.

Commit:
Fixing bug with suggestions visibility

Commit:
Fixing bug with suggestions visibility

Commit:
Setting suggestions as an empty array by default

toranb commented:
It looks like travis blew up -any chance you could give this a look before we merge it in?

anthonygore commented:
Sure, I'll have another look. Sorry for the delay.

Commit:
onInput bug

Commit:
onInput bug

Commit:
Updating included function so this addon can be used within another addon

anthonygore commented:
Hi Toran,

I've got this sorted out now. Probably best to ignore what I wrote above and let me explain the changes again from scratch!

Comparing my latest commit to your repo: https://github.com/toranb/ember-cli-auto-complete/compare/master...anthonygore:master

1) Changes to `README.md`

The documentation currently says that you need to implement a function `suggestions`. It should actually be `determineSuggestions`. I made that change in my last PR and it was wrong.

2) Changes to `auto-complete.js`

I've changed two lines here to fix a bug. The use case that shows the bug is this: if you have a really long list of suggestions, you may not want to return any suggestions until the user has typed at least n characters so as not to return a huge array. The bug is that the suggestions list stays hidden until the user has typed n+1 characters, not n. For example, lets say you have 20,000 items in your suggestion list, and you decide you won't return anything until the user types at least 3 characters, because that reduces the suggestion array to just 50 records. The existing code will return a suggestion array based on a 3 character match only when `inputVal.length` is 4. So these changes fix that.

3) Changes to `index.js`

The `included` hook breaks when you nest ember-cli-auto-correct within another addon. This issue is documented here https://github.com/ember-cli/ember-cli/issues/3718 and I've implemented the same solution from ember-cli-pretender https://github.com/rwjblue/ember-cli-pretender/commit/dda9bf88597e5f127cfbe0cf863f2b03c4e24246

toranb commented:
Wow thanks for the excellent work on this @anthonygore !!! The only item I'll mention is that 3) won't be an issue for long as I plan to pull the css from this before we ship v1.0.0 offically. I started by pushing the css down but found that most people don't use this "global css" approach and it can be harmful so ... I decided a big part of the breaking change to v1.0 would be to not pass along the vendor/css
